### PR TITLE
Add native polygonstolines algorithm

### PR DESCRIPTION
**Pull request by m-kuhn**
A 1:1 port of the existing Python polygons to lines algorithm.
Side product of a very confusing sip memory management debug session.

* [x] Unit Tests

### CODE REVIEW

**Comment by nirvn** [comment]
@m-kuhn , did you ever find what was wrong at the end?

File: src/analysis/processing/qgsalgorithmpolygonstolines.cpp [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+  if ( QgsGeometryCollection *collection = qgsgeometry_cast<QgsGeometryCollection *>( geom ) )
+  {
+    QgsGeometryPartIterator parts = collection->parts();
+    while ( parts.hasNext() )
+      rings.append( extractRings( parts.next() ) );
+  }
+  else if ( QgsCurvePolygon *polygon = qgsgeometry_cast<QgsCurvePolygon *>( geom ) )
+  {
+    rings.append( polygon->exteriorRing()->clone() );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmpolygonstolines.cpp]: Do we need to guard against null exterior ring ? (i think we should).

**Comment by nyalldawson** [comment]
I'm definitely in favour of python -> C++ ports!

**Comment by m-kuhn** [comment]
> @m-kuhn , did you ever find what was wrong at the end?

See here: https://www.riverbankcomputing.com/pipermail/pyqt/2019-January/041251.html

The most probable suspicion is that a wrapper around a QgsAbstractGeometry (QgsPoint) pointer is cached in some Python code while the underlying object gets deleted and a new one (QgsMultiPolygon) created at the same address and the existing (QgsPoint) wrapper reused.

Conclusion: never cache a QgsAbstractGeometry pointer in Python, always cache the complete QgsGeometry (or clone() the QgsAbstractGeometry but I don't see the benefit of doing that).

**Comment by m-kuhn** [comment]
Not really, it's just a guideline python devs need to follow

File: src/analysis/processing/qgsalgorithmpolygonstolines.cpp (hunk 1 above)
Comment (m-kuhn) [reply src/analysis/processing/qgsalgorithmpolygonstolines.cpp]: Also added a test for that

**Comment by m-kuhn** [comment]
Added some boilerplate code for quick and dirty feature based algorithm tests
Includes a forward port of #8913

**Comment by nirvn** [comment]
I'd say +1 for 3.6; straightforward porting to c++ to fix a SIP issue is bug fixing. The extra benefits are just a cherry on top of the fix :)

**Comment by nyalldawson** [comment]
+1 to 3.6

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
@m-kuhn 
I'm not sure Phil's interpretation of the issue is correct here. See https://issues.qgis.org/issues/20661#note-2 for another test case, which definitely **isn't** being cached.

**Comment by m-kuhn** [comment]
>  I'm not sure Phil's interpretation of the issue is correct here. See https://issues.qgis.org/issues/20661#note-2 for another test case, which definitely isn't being cached.

Ouch, not sure what to make of that...

**Pull request merged**